### PR TITLE
RUM-1197 fix: Context Propagation on FeatureBaggage reset

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Setting `nil` for a Feature baggage in core would store a `FeatureBaggage(nil)` instead of removing it from the context. 

```swift
let rum: RUMContext? = nil
core.set(baggage: rum, forKey: "rum")
// ...
print(context.baggages["rum"] == nil) // ❌ output: "false"
```

### How?

Add support for nillifying a `baggage` from the `DatadogCoreProtocol` interface and add test coverage on all feature that can access a `nil` or malformed baggage.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests for Core, RUM, Trace, Logs, CR and WVT
- [x] Run unit tests for Session Replay
- [x] Run integration tests
- [ ] Run smoke tests
- [ ] Run tests for `tools/`

### CODE REVIEW

File: DatadogTrace/Sources/Feature/MessageReceivers.swift [hunk 1]
```diff
@@ -65,6 +65,8 @@ internal final class ContextMessageReceiver: FeatureMessageReceiver {
                     core.telemetry
                         .error("Fails to decode RUM context from Trace", error: error)
                 }
+            } else {
+                $0.rum = nil
```
Comment (maxep) [inline DatadogTrace/Sources/Feature/MessageReceivers.swift:69]: We were not resetting the RUM context for Trace. **This flaw has not been released.**

File: DatadogSessionReplay/Tests/Feature/RUMContextReceiverTests.swift [hunk 2]
```diff
@@ -115,18 +95,69 @@ class RUMContextReceiverTests: XCTestCase {
         XCTAssertEqual(rumContexts[1].viewServerTimeOffset, 345)
     }
 
-    func testWhenMessageDoesntContainRUMBaggage_itCallsFallback() {
-        let context = DatadogContext.mockAny()
-        let message = FeatureMessage.context(context)
```
Comment (ncreated) [inline DatadogSessionReplay/Tests/Feature/RUMContextReceiverTests.swift:120]: blocking/ With this removal, we're no longer covering the behaviour of `RUMContextReceiver` acknowledging the "context message" - I mean [this line](https://github.com/DataDog/dd-sdk-ios/blob/101fe09b2dd1d444e7cca262874a5ffa3396ea40/DatadogSessionReplay/Sources/Feature/RUMContextReceiver.swift#L60).

These tests seem to suffer the same problem as `WebViewLogReceiverTests` - what we test is integration with mocked core, instead of testing the standalone receiver.

File: DatadogTrace/Tests/ContextMessageReceiverTests.swift [hunk 3]
```diff
@@ -67,6 +68,56 @@ class ContextMessageReceiverTests: XCTestCase {
...
+        XCTAssertEqual(receiver.context.rum?["_dd.session.id"], "session-id")
+        XCTAssertEqual(receiver.context.rum?["_dd.view.id"], "view-id")
+        XCTAssertEqual(receiver.context.rum?["_dd.action.id"], "action-id")
+
+        // When
+        core.set(baggage: nil, forKey: "rum")
+
+        // Then
+        XCTAssertNil(receiver.context.rum)
+    }
```
Comment (ncreated) [inline DatadogTrace/Tests/ContextMessageReceiverTests.swift:98]: Similar problem to other `MessageReceiver` tests. We test the receiver's integration with **mocked** core, not the receiver itself. The same in `testItReceivesMalformedRUMContext()`.

**Review by ncreated — CHANGES_REQUESTED** [review]

File: DatadogInternal/Tests/DatadogCoreProtocolTests.swift [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        let core = CoreMock()
+
+        // When
+        core.send(message: .baggage(key: "test", value: "value"))
+
+        // Then
+        XCTAssertEqual(
+            try core.message?.baggage(forKey: "test"), "value", "DatadogCoreProtocol.send(message:) should forward message"
+        )
+    }
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogInternal/Tests/DatadogCoreProtocolTests.swift:41]: This test is not testing the `FeatureBaggage` encoding/decoding but the `DatadogCoreProtocol` extension, asserting that the `core.send` properly forward the message. I could test other types of message tho, and I'm happy to add more fuzzy inputs to `FeatureBaggageTests`!

File: DatadogLogs/Tests/WebViewLogReceiverTests.swift [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import XCTest
+import TestUtilities
+import DatadogInternal
+
+@testable import DatadogLogs
+
+class WebViewLogReceiverTests: XCTestCase {
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogLogs/Tests/WebViewLogReceiverTests.swift:13]: I'm sorry but I don't see the violation? It uses the `PassthroughCoreMock` because we are in the module unit tests, and the `WebViewLogReceiver` is set as a receiver of the core. The `WebViewLogReceiver` has no other API that `receive`. I could do the following:
```swift
messageReceiver.receive(
    message: .baggage(...),
    from: core
)
```
But that is essentially the same, no?

[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogLogs/Tests/WebViewLogReceiverTests.swift:13]: It is totally fine to use `PassthroughCoreMock` 👍 as we don't care about the real core behaviour in these tests. For the same reason, yes we should test:
```swift
messageReceiver.receive(message:from:)
```
and not `.send(message:else:)` on core mock. 

Otherwise we cannot test all behaviours without re-implementing real logic into `PassthroughCoreMock`. Take an example: _"given message addressed to other feature, web view log receiver rejects it"_ - it is a matter of asserting `false` returned from `receive(message:from:)` vs basing assertion on [re-implemented fallback logic](https://github.com/DataDog/dd-sdk-ios/blob/336d1dcf78402e7a00310a5b081007aa7210a344/TestUtilities/Mocks/CoreMocks/PassthroughCoreMock.swift#L108) in mock.

Exactly this change of paradigm was the actual thing that surfaced the `FeatureBaggage` propagation problems in https://github.com/DataDog/dd-sdk-ios/pull/1502/files#r1341160540. For that reason, I'm advocating for less dependency on mocks, more edge case coverage and more fuzzy testing.

[the remark this answers is not in the document]
Comment (maxep) [reply DatadogLogs/Tests/WebViewLogReceiverTests.swift:13]: Ok, yes it makes sense! Thanks for the clarification 🙏

File: DatadogInternal/Tests/DatadogCoreProtocolTests.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogInternal/Tests/DatadogCoreProtocolTests.swift]: Yes, fuzzy input would make more sense in `FeatureBaggageTests` than here 👍. We can revisit it with next occasion.
